src/sinks/datadog/metrics/sink.rs

```diff
@@ -145,3 +154,270 @@ where
+                    // Collapse this counter by accumulating its value, and its
+                    // finalizers, and removing it from the original vector of metrics.
+                    accumulated_value += *value;
+
+                    let mut old_metric = metrics.swap_remove(inner_idx);
+                    accumulated_finalizers.merge(old_metric.metadata_mut().take_finalizers());
+                    should_advance = false;
+                } else {
+                    // We hit a counter that _doesn't_ match, but we can't just skip
+                    // it because we also need to evaulate it against all the
+                    // counters that come after it, so we only increment the index
+                    // for this inner loop.
+                    //
+                    // As well, we mark ourselves to stop incrementing the outer
+                    // index if we find more counters to accumulate, because we've
+                    // hit a disjoint counter here. While we may be continuing to
+                    // shrink the count of remaining metrics from accumulating,
+                    // we have to ensure this counter we just visited is visited by
+                    // the outer loop.
+                    is_disjoint = true;
+                }
+            }
+
+            if should_advance {
+                inner_idx += 1;
+
+                if !is_disjoint {
+                    idx += 1;
+                }
+            }
+        }
+
+        // If we had matches during the accumulator phase, update our original counter.
+        if had_match {
+            let metric = metrics.get_mut(curr_idx).expect("current index must exist");
+            match metric.value_mut() {
+                MetricValue::Counter { value } => {
+                    *value += accumulated_value;
+                    metric
+                        .metadata_mut()
+                        .merge_finalizers(accumulated_finalizers);
+                }
+                _ => unreachable!("current index must represent a counter"),
+            }
+        }
+
+        idx += 1;
+    }
+
+    metrics
+}
+
+#[cfg(test)]
+mod tests {
+    use chrono::{DateTime, Utc};
+    use proptest::prelude::*;
+    use vector_core::event::{Metric, MetricKind, MetricValue};
+
+    use super::collapse_counters_by_series_and_timestamp;
+
+    fn arb_collapsible_metrics() -> impl Strategy<Value = Vec<Metric>> {
+        let ts = Utc::now();
+
+        any::<Vec<(u16, MetricValue)>>().prop_map(move |values| {
+            values
+                .into_iter()
+                .map(|(id, value)| {
+                    let name = format!("{}-{}", value.as_name(), id);
+                    Metric::new(name, MetricKind::Incremental, value).with_timestamp(Some(ts))
+                })
+                .collect()
+        })
+    }
+
+    fn create_counter(name: &str, value: f64) -> Metric {
+        Metric::new(
+            name,
+            MetricKind::Incremental,
```

Comment:
Would it be worth testing handling of absolute counters, or is that not relevant here?

Reply:
We always force counters to be incremental for the `datadog_metrics` sink in particular, which is why these tests elide testing both variants.